src/cats/CatSlideshow.js

```diff
@@ -0,0 +1,219 @@
+import usePaginatedFetch from "./usePaginatedFetch";
+import CatSlideshowControls from "./CatSlideshowControls.js";
+import { SlideAnimation, Direction } from "./SlideAnimation.js";
+import { LoadingCard } from "./Cards.js";
+import "./CatSlideshow.css";
+
+const PAGE_SIZE = 20;
+const PAGE_PREFETCH = 8;
+const MAX_PREFETCH = 2;
+
+function fetchPageFromCatsApi(pageIndex, selectedBreedID, fetchPage) {
+  const { url, options } = getCatsApiFetchParams("/images/search", {
+    breed_id: selectedBreedID === "all" ? "" : selectedBreedID,
+    limit: PAGE_SIZE,
+    order: "ASC",
+    page: pageIndex,
+  });
+
+  const getPageData = async (response) => {
+    const json = await response.json();
+    return json.map((item) => ({ id: item.id, url: item.url }));
+  };
+
+  const getMetadata = (response) => {
+    return Object.fromEntries(
+      [...response.headers.entries()].filter(
+        (entry) => entry[0] === "pagination-count"
+      )
+    );
+  };
+
+  fetchPage(url, options, pageIndex, selectedBreedID, getPageData, getMetadata);
+}
+
+function getPageIndex(index) {
+  return Math.floor(index / PAGE_SIZE);
+}
+
+function getImageURL(pages, index) {
+  const pageIndex = getPageIndex(index);
+  const offset = index % PAGE_SIZE;
+  return pages?.[pageIndex]?.data?.[offset]?.url;
+}
+
+function initCatSlideshowState() {
+  return {
+    direction: Direction.Next,
+    index: 0,
+    maxIndex: null,
+    visibleIndex: null,
+    readyMap: {},
+  };
+}
+
+function catSlideshowReducer(state, action) {
+  //console.log("%ccat slideshow: " + action.type, "color: red", action, state);
+
+  function changeIndex(state, increment) {
+    const newIndex = Math.min(
+      Math.max(0, state.index + increment),
+      state.maxIndex
+    );
+    const isReady = state.indexReadyMap[newIndex] === "ready";
+
+    return {
+      ...state,
+      direction: increment > 0 ? Direction.Next : Direction.Previous,
+      index: newIndex,
+      visibleIndex: isReady ? newIndex : state.visibleIndex,
+    };
+  }
+
+  switch (action.type) {
+    case "change-breed":
+      return initCatSlideshowState();
+    case "update-max-index":
+      return { ...state, maxIndex: action.maxIndex };
+    case "decrement-index":
+      return changeIndex(state, -1);
+    case "increment-index": {
+      return changeIndex(state, 1);
+    }
+    case "prefetch-status": {
+      const newState = {
+        ...state,
+        indexReadyMap: { ...state.indexReadyMap, [action.index]: "ready" },
+      };
+
+      if (action.index === state.index) {
+        newState.visibleIndex = action.index;
+      }
+
+      return newState;
+    }
+    default:
+      throw new Error(`Unhandled action type: ${action.type}`);
+  }
+}
+
+function CatSlideshow({ selectedBreedID }) {
+  const prevSelectedBreedIDRef = useRef(selectedBreedID);
+  const prefetchMapRef = useRef(new Map());
+  const { pages, metadata, fetchPage, resetPages } = usePaginatedFetch(
```

Comment:
Sure. I will make those changes and see how the code works out. I think I learned from my studies that reducers have some really solid properties (stable dispatch function, clarity, testability) and so I tended to use them all over the place. 

What's your decision process on when to use a reducer?
1) If you need to have logic that updates multiple pieces of dependent state
2) when you want to avoid passing several callback props to a child
3) If you end up with duplicated state update logic in your component
4) ...?

Reply:
So I pulled out the two reducers, as you suggest. It made a huge difference in clarity to usePaginatedFetch and it's clearly better. It also reminded me that the functional version of setState is super useful inside async responses so you don't get bit by stale state.

It did also improve CatSlideshow but the difference is less marked. For components that need to reset all their state occasionally, one function can be used to initialize the reducer and also reset the state (see line 48). After replacing the reducer with 5 useState calls, the reset function feels somewhat duplicative. Do you have a suggestion on how to improve that?

Reply:
Yes, `useReducer` is definitely good and you found good resources on that. I used it recently to drive the entire state of a virtualized list and it worked out really well and was highly testable.